src/Lean/Meta/Tactic/Simp/Types.lean

```diff
@@ -97,6 +99,19 @@ def getSimpCongrTheorems : M SimpCongrTheorems :=
   finally
     modify fun s => { s with cache := cacheSaved }
 
+def recordSimpTheorem (n : Name) : SimpM Unit :=
+  modify fun s => { s with usedTheorems := s.usedTheorems.insert n }
+
+def ppSimpTheorem (s : SimpTheorem) : SimpM MessageData :=
+  s.toMessageDataM fun n => do
+    if (← getEnv).contains n then
+      return ← mkConstWithLevelParams n
+    if (← getLCtx).contains ⟨n⟩ then
+      return mkFVar ⟨n⟩
+    if let some stx := (← read).namedStx.find? n then
+      return stx
+    return n
```

Comment:
This `n` here is the weird part. The user may see the free variable identifiers. I suspect it can happen if we are using `contextual := true`, and we have something like

```
if x = 0 then 
 ... f x ... -- `x` is going to be rewritten here, and the theorem used will not be in the local context.
else
  ....
```

Reply:
I found your comment at `traceSimpCall`. So, you are aware of this. I am now assuming this code will never be invoked to trace a theorem if `n` is the free variable id of a theorem that is not in scope anymore.